src/openrct2-ui/ride/VehicleSounds.cpp

```diff
@@ -0,0 +1,608 @@
+#include "VehicleSounds.h"
+
+#include "../interface/Viewport.h"
+#include "../interface/Window.h"
+
+#include <openrct2/Context.h>
+#include <openrct2/GameState.h>
+#include <openrct2/OpenRCT2.h>
+#include <openrct2/audio/AudioChannel.h>
+#include <openrct2/audio/AudioMixer.h>
+#include <openrct2/audio/audio.h>
+#include <openrct2/entity/EntityRegistry.h>
+#include <openrct2/profiling/Profiling.h>
+#include <openrct2/ride/TrainManager.h>
+#include <openrct2/ride/Vehicle.h>
+
+namespace OpenRCT2::Audio
+{
+    namespace
+    {
+        template<typename T> class TrainIterator;
+        template<typename T> class Train
+        {
+        public:
+            explicit Train(T* vehicle)
+                : FirstCar(vehicle)
+            {
+                assert(FirstCar->IsHead());
+            }
+            int32_t Mass();
+
+            friend class TrainIterator<T>;
+            using iterator = TrainIterator<T>;
+            iterator begin()
+            {
+                return iterator{ FirstCar };
+            }
+            iterator end()
+            {
+                return iterator{};
+            }
+
+        private:
+            T* FirstCar;
+        };
+        template<typename T> class TrainIterator
+        {
+        public:
+            using iterator = TrainIterator;
+            using iterator_category = std::forward_iterator_tag;
+            using value_type = T;
+            using pointer = T*;
+            using reference = T&;
+
+            TrainIterator() = default;
+            explicit TrainIterator(T* vehicle)
+                : Current(vehicle)
+            {
+            }
+            reference operator*()
+            {
+                return *Current;
+            }
+            iterator& operator++()
+            {
+                Current = GetEntity<Vehicle>(NextVehicleId);
+                if (Current != nullptr)
+                {
+                    NextVehicleId = Current->next_vehicle_on_train;
+                }
+                return *this;
+            }
+            iterator operator++(int)
+            {
+                iterator temp = *this;
+                ++*this;
+                return temp;
+            }
+            bool operator!=(const iterator& other)
+            {
+                return Current != other.Current;
+            }
+
+        private:
+            T* Current = nullptr;
+            EntityId NextVehicleId = EntityId::GetNull();
+        };
+    } // namespace
+
+    template<typename T> int32_t Train<T>::Mass()
+    {
+        int32_t totalMass = 0;
+        for (const auto& vehicle : *this)
```

Comment:
Since we have iterators we should use std::accumulate, we also have C++20 so ranges::accumulate.

Reply:
Eh, well std::accumulate then.